Finally tell the wizard story (#14291)
PostHog/posthog.com: contents/blog/correct-llm-code-generation.mdx

---
title: "You can have LLM code that's actually good"
date: 2026-01-06
author:
  - danilo-campos
featuredImage: >-
    https://res.cloudinary.com/dmukukwp6/image/upload/robot_6d2cab1b66.png
featuredImageType: full
tags:
    - Engineering
    - LLM
---

import Prompt from '../docs/integrate/_snippets/agent-prompt.mdx'

You know what’s boring?

Writing integrations.

> It’s a task you’ll do once, but you always need to do it right, which means carefully reading docs. Depending on your project’s framework, there might be some caveats you need to think about.

Reviewer: ```suggestion
It’s a task you’ll do once, but you always need to do it right, which means carefully reading docs. Since you're working with an unfamiliar library, there are always caveats you need to think about.
```

I'd lean into it more. I _rarely_ get it right the first time with any SDK, heck I screw up with system libs lol.

 
All of this sucks. So we built the PostHog AI Wizard: a command line tool that tailors a correct integration of our product for you.

This tool can understand your business goals, where churn happens, and the conditions for your success, all as expressed in your code.

And it can do this with just one command.

![Wizard run](https://res.cloudinary.com/dmukukwp6/image/upload/wizard_run_adb8660e2a.gif)

<Prompt />

> Try it out with your [Next.js projects](https://posthog.com/docs/libraries/next-js). We’re adding more frameworks and languages as you read this.

Reviewer 2: Feels like this CTA belongs between the gif and prompt. 
Then "I spent...." is the start of a new thought.

You can use LLMs to generate correct and predictable code, too. Here’s everything we learned. 

## The dead-end of v1

Our [first version](/blog/envoy-wizard-llm-agent) made you happy and we love that.

But architecturally, it was a dead-end.

Maintaining it was a bear: documentation for the LLM was embedded in its code. Because of how fast both software frameworks and PostHog itself evolves, the documentation was constantly falling out of sync with our website, not the mention the web frameworks it targeted.

This was a great MVP, but it emerged before we were all savvy to [context engineering](https://www.anthropic.com/engineering/effective-context-engineering-for-ai-agents)

![Wizard v1](https://res.cloudinary.com/dmukukwp6/image/upload/q_auto,f_auto/run_pray_finish_6daa2fea52.png)

Worse still, the Wizard relied on a sort of Pareto luck: 80% of the time it could make the right edits. The rest of the time, it completely failed. It was a single-shot edit driven by an LLM, but scaffolded by a bunch of conventional code that hoped to find the right files. If your project was just a little weird, even just part of a monorepo, you were out of luck.

> Most damning, the Wizard didn’t know if it succeeded. It crossed its fingers.

Reviewer: I really wish the wizard had a name

Norbert Wiener, father of cybernetics, [writes](https://en.wikipedia.org/wiki/The_Human_Use_of_Human_Beings):

> the structure of the machine or organism is an index of the performance that may be expected from it.

This observation, echoing 75 years from the past, *rang true*: we had a CLI that was built to explore a specific set of paths that might contain relevant files. It could read them and write them. But this was a straitjacket. We couldn’t possibly code ahead of all the permutations that it might encounter.

Its performance was capped by its structure.

So we needed new structure.

Instead of a car on a rail, we needed something that could steer, even backtrack, steadily circling in on its goal.

A tool like the [Claude Agent SDK](https://platform.claude.com/docs/en/agent-sdk/overview) is perfect for this: it’s structured to both explore and integrate feedback. If a file isn’t where an agent expects, it can run a bash command to go looking for it.

By contrast to our crude CLI, the agent was nimble as a spider. It could detect coding errors, and it could trigger linters, type checkers, and prettiers. It could come at a problem from multiple angles.

> ![Wizard v2](https://res.cloudinary.com/dmukukwp6/image/upload/q_auto,f_auto/wizardv2_8a75a8be43.png)

Reviewer 2: Nice 👌

The agent had all of the information encoded in the language model to inform its moves, and a broad set of tools it could use to make those moves.

Agents need less luck. But they do need loads of context.

> 

Reviewer 3: ```suggestion
```

Reviewer 3: i hate this this extra line bothers me

This, understandably, made folks uneasy. *Did the robot hallucinate this pattern?*

The future of the Wizard needed to be hand-in-glove with whatever we’d written on posthog.com. Both sources had to agree.

Today, the Wizard is always primed with up-to-the-second documentation from the website. If a framework gets a new way of integrating analytics, all we need is a website update for both humans and the wizard to do better. This saves a ton of labor, and it lets us trim back the wizard to something lighter. No more docs embedded in the code.

But prose documentation alone, it turned out, was not enough. We needed a little more detail to ensure correctness. So for every framework we support, we build a [toy project](https://github.com/PostHog/examples/tree/main/basics) in code. These are not starter projects and you should not ship them: most obviously, the authorization flows are completely fake. They accept any password. Still, they demonstrate a PostHog integration in motion, giving the agent richer influence for doing the right thing.

## Progressive disclosure

> The last wrinkle for a robot making thousands of project edits per month was harder: how do you do the work *predictably?*

Reviewer 2: This, and the earlier mention of **robot** tingle my brain in a weird way. My notion of a robot is embodied AI, or the very least a physical machine. Since the post is about LLMs, the semantics might actually matter here.

LLMs are non-deterministic systems. That’s the devil’s agreement of working with them. An agent can come up with many patterns and sequences for accomplishing the same goal.

What worked best for us was bread-crumbing the prompts, revealing them in sequence. Instead of dumping the entire plan right away, we disclose bits at a time.

> You can do this without any fancy agent orchestration.

Reviewer 3: i feel like you can even be more explicit: the answer wasn't code, it was context. words!
